tempesta_fw/http.c

```diff
@@ -382,6 +370,396 @@ tfw_http_send_504(TfwHttpReq *req)
 	return tfw_http_send_resp(req, &rh, __TFW_STR_CH(&rh, 1));
 }
 
+static inline bool
+tfw_http_req_is_nonidempotent(TfwHttpReq *req)
+{
+	return (req->flags & TFW_HTTP_NON_IDEMP);
+}
+
+/*
+ * Set the request @req in server connection @srv_conn as idempotent.
+ * Called only when a request turns idempotent from a non-idempotent.
+ */
+static inline void
+__tfw_http_req_nonidemp_delist(TfwConnection *srv_conn, TfwHttpReq *req)
+{
+	list_del_init(&req->nip_list);
+	if (list_empty(&srv_conn->nip_queue))
+		clear_bit(TFW_CONN_B_HASNIP, &srv_conn->flags);
+}
+
+/*
+ * Set the request @req in server connection @srv_conn as non-idempotent.
+ */
+static inline void
+__tfw_http_req_nonidemp_enlist(TfwConnection *srv_conn, TfwHttpReq *req)
+{
+	BUG_ON(!list_empty(&req->nip_list));
+	list_add_tail(&req->nip_list, &srv_conn->nip_queue);
+	set_bit(TFW_CONN_B_HASNIP, &srv_conn->flags);
+}
+
+/*
+ * If @req in server connection @srv_conn is non-idempotent, then set
+ * it as idempotent.
+ */
+static inline void
+tfw_http_req_nonidemp_delist(TfwConnection *srv_conn, TfwHttpReq *req)
+{
+	if (!list_empty(&req->nip_list))
+		__tfw_http_req_nonidemp_delist(srv_conn, req);
+}
+
+/*
+ * If a request on the list of non-idempotent requests in server
+ * connection @srv_conn had become idempotent, then set it as such.
+ */
+static inline void
+tfw_http_conn_nonidemp_delist(TfwConnection *srv_conn)
+{
+	TfwHttpReq *req, *tmp;
+
+	list_for_each_entry_safe(req, tmp, &srv_conn->nip_queue, nip_list)
+		if (!tfw_http_req_is_nonidempotent(req)) {
+			BUG_ON(list_empty(&req->nip_list));
+			__tfw_http_req_nonidemp_delist(srv_conn, req);
+		}
+}
+
+/*
+ * Tell if the server connection's forwarding queue is on hold.
+ * It's on hold it the request that was sent last was non-idempotent.
+ */
+static inline bool
+tfw_http_conn_on_hold(TfwConnection *srv_conn)
+{
+	TfwHttpReq *req = (TfwHttpReq *)srv_conn->msg_sent;
+
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+	return (req && tfw_http_req_is_nonidempotent(req));
```

Comment:
This particular function specifically checks if it's allowed to continue forwarding. So the checks are not equivalent or duplicate.

The presence of non-idempotent requests in the connection does not mean we have to stop forwarding. Those requests may be somewhere farther in the queue, and the forwarding may continue until one of those is sent out. When a non-idempotent requests is sent out (`conn->req_sent` is non-idempotent), then forwarding must be paused until a response is received to that request.